dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java

```diff
@@ -41,26 +43,30 @@
     @Nullable
     private static final Void NO_CONTEXT = null;
 
-    private static final Ticker SYSTEM_NANOTIME = System::nanoTime;
-
-    private final Meter limitedMeter;
     private final LimitedChannel delegate;
-    private final Limiter<Void> limiter;
+    private final SimpleLimiter<Void> limiter;
+    private final Instrumentation instrumentation;
 
-    @VisibleForTesting
-    ConcurrencyLimitedChannel(LimitedChannel delegate, Limiter<Void> limiter, DialogueClientMetrics metrics) {
+    ConcurrencyLimitedChannel(LimitedChannel delegate, SimpleLimiter<Void> limiter, Instrumentation instrumentation) {
         this.delegate = new NeverThrowLimitedChannel(delegate);
-        this.limitedMeter = metrics.limited(getClass().getSimpleName());
+        this.instrumentation = instrumentation;
         this.limiter = limiter;
+        instrumentation.registerMaxGauge(limiter::getLimit);
+        instrumentation.registerUtilizationGauge(this::getUtilization);
     }
 
-    static ConcurrencyLimitedChannel create(LimitedChannel delegate, DialogueClientMetrics metrics) {
-        return new ConcurrencyLimitedChannel(
-                delegate, ConcurrencyLimitedChannel.createLimiter(SYSTEM_NANOTIME), metrics);
+    interface Instrumentation {
+        /** Called every time this channel is unable to process a request due to concurrency limits. */
+        void markLimited();
+
+        /** Called exactly once to register a gauge. */
+        void registerUtilizationGauge(DoubleSupplier gauge);
+
+        /** Called exactly once to register a gauge. */
+        void registerMaxGauge(IntSupplier gauge);
     }
 
-    @VisibleForTesting
-    static Limiter<Void> createLimiter(Ticker nanoTimeClock) {
+    static SimpleLimiter<Void> createLimiter(Ticker nanoTimeClock) {
```

Comment:
could we wrap this in instrumentation instead?

Reply:
I've just tried this (see snippet), but we'd end up jumping through so many hoops to try and avoid gauges (counters only support integers, whereas utilization is a ratio), they also don't allow us to just set a new value.

```java
    private static class InstrumentedSimpleLimiter implements Limiter<Void> {
        private final SimpleLimiter<Void> simpleLimiter;

        private InstrumentedSimpleLimiter(
                SimpleLimiter<Void> simpleLimiter, DialogueConcurrencylimiterMetrics metrics, int hostIndex) {
            this.simpleLimiter = simpleLimiter;
        }

        private void updateMetrics(Instrumentation instrumentation, SimpleLimiter<Void> simpleLimiter) {
            instrumentation.updateMax(simpleLimiter.getLimit());
            instrumentation.updateUtilization(((double) simpleLimiter.getInflight()) / simpleLimiter.getLimit());
        }

        @Override
        public Optional<Listener> acquire(Void context) {
            Optional<Listener> maybeListener = simpleLimiter.acquire(context);
            if (!maybeListener.isPresent()) {
                return Optional.empty();
            }
            return Optional.of(new Listener() {
                @Override
                public void onSuccess() {
                    maybeListener.get().onSuccess();
                    updateMetrics(instrumentation, simpleLimiter);
                }

                @Override
                public void onDropped() {
                    maybeListener.get().onDropped();
                    updateMetrics(instrumentation, simpleLimiter);
                }

                @Override
                public void onIgnore() {
                    maybeListener.get().onIgnore();
                }
            });
        }
    }
```

Reply:
ok if its not coming out cleanly we can try the gauges